File: leetcode/easy/cpp/LC_3407.cpp
#include <iostream>
#include <string>

using namespace std;

class Solution {
public:
    bool hasMatch(string s, string p) {
        int n = s.length();

        size_t pos = p.find('*');
        string prefix = p.substr(0, pos);
        string sufix = p.substr(pos+1);

        int prefix_len = prefix.length();
        int sufix_len = sufix.length();

        bool prefix_found = (prefix_len == 0);
        bool sufix_found = (sufix_len == 0);

        int idx = 0;

        while(!prefix_found && idx <= n-prefix_len) {
            if(s.substr(idx, prefix_len) == prefix) {
                prefix_found = true;
                idx += prefix_len;
            } else {
                idx++;
            }
        }

        while(prefix_found && idx <= n-sufix_len) {
            if(s.substr(idx, sufix_len) == sufix) {
                sufix_found = true;
                break;
            }
            idx++;
        }
        return prefix_found && sufix_found;
    }
};

int main() {
    Solution solution;
    string s, p;
    cin >> s;
    cin >> p;
    cout << (solution.hasMatch(s=s, p=p)? "true": "false");
    return 0;
}

File: leetcode/easy/python/LC_3407.py
class Solution:
    def hasMatch(self, s: str, p: str) -> bool:
        n = len(s)

        parts = p.split("*")
        prefix, sufix = parts[0], parts[1]

        prefix_len, sufix_len = len(prefix), len(sufix)

        prefix_found = True if prefix_len == 0 else False
        sufix_found = True if sufix_len == 0 else False

        idx = 0

        while not prefix_found and idx <= n-prefix_len:
            if s[idx: idx+prefix_len] == prefix:
                prefix_found = True
                idx += idx+prefix_len
            else:
                idx += 1

        while prefix_found and idx <= n-sufix_len:
            if s[idx: idx+sufix_len] == sufix:
                sufix_found = True
                break
            idx += 1
        
        return prefix_found and sufix_found

solution = Solution()
s = input()
p = input()
print("true" if solution.hasMatch(s=s, p=p) else "false")